File: src/main/java/com/github/zametki/db/cache/ZametkaCache.java
package com.github.zametki.db.cache;

import com.github.zametki.model.UserId;
import com.github.zametki.model.Zametka;
import com.github.zametki.model.ZametkaId;
import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

public class ZametkaCache {

    private final Map<ZametkaId, Zametka> zametkaById = new ConcurrentHashMap<>();

    private final Map<UserId, List<ZametkaId>> zametkaByUser = new ConcurrentHashMap<>();

    public Zametka getById(@NotNull ZametkaId zametkaId, @NotNull Function<ZametkaId, Zametka> loader) {
        return zametkaById.computeIfAbsent(zametkaId, loader);
    }

    public void update(@NotNull Zametka zametka) {
        zametkaById.put(zametka.id, zametka);
    }

    public void remove(@NotNull ZametkaId zametkaId) {
        zametkaById.remove(zametkaId);
    }

    public List<ZametkaId> getByUser(@NotNull UserId userId, @NotNull Function<UserId, List<ZametkaId>> loader) {
        return zametkaByUser.computeIfAbsent(userId, id -> Collections.unmodifiableList(loader.apply(id)));
    }

    public void remove(@NotNull UserId userId) {
        zametkaByUser.remove(userId);
    }
}

File: src/main/java/com/github/zametki/db/cache/ZametkaCache.kt
package com.github.zametki.db.cache

import com.github.zametki.model.UserId
import com.github.zametki.model.Zametka
import com.github.zametki.model.ZametkaId
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Function

class ZametkaCache {

    private val zametkaById = ConcurrentHashMap<ZametkaId, Zametka?>()

    private val zametkaByUser = ConcurrentHashMap<UserId, List<ZametkaId>>()

    fun getById(zametkaId: ZametkaId, loader: Function<ZametkaId, Zametka?>) = zametkaById.computeIfAbsent(zametkaId, loader)

    fun update(zametka: Zametka) {
        zametkaById.put(zametka.id!!, zametka)
    }

    fun remove(zametkaId: ZametkaId) {
        zametkaById.remove(zametkaId)
    }

    fun getByUser(userId: UserId, loader: Function<UserId, List<ZametkaId>>) = zametkaByUser.computeIfAbsent(userId) { id -> Collections.unmodifiableList(loader.apply(id)) }

    fun remove(userId: UserId) {
        zametkaByUser.remove(userId)
    }
}
